ArduCopter/Copter.cpp

```diff
@@ -389,7 +389,59 @@ bool Copter::set_target_angle_and_climbrate(float roll_deg, float pitch_deg, flo
     mode_guided.set_angle(q, Vector3f{}, climb_rate_ms*100, false);
     return true;
 }
-#endif
+
+bool Copter::set_target_rate_and_throttle(float roll_rate_dps, float pitch_rate_dps, float yaw_rate_dps, float throttle)
+{
+    // exit if vehicle is not in Guided mode or Auto-Guided mode
+    if (!flightmode->in_guided_mode()) {
+        return false;
+    }
+
+    // Zero quaternion indicates rate control
+    Quaternion q;
+    q.zero();
+
+    // Convert from degrees per second to radians per second
+    Vector3f ang_vel_body { roll_rate_dps, pitch_rate_dps, yaw_rate_dps };
+    ang_vel_body *= DEG_TO_RAD;
+
+    // Pass to guided mode
+    mode_guided.set_angle(q, ang_vel_body, throttle, true);
+    return true;
+}
+
+// Register a custom mode with given number and names
+bool Copter::register_custom_mode(const uint8_t num, const char* full_name, const char* short_name)
+{
+    // Numbers over 100 are reserved for custom modes
+    if (num < 100) {
```

Comment:
#define somewhere for this magic number

Reply:
I don't like defining stuff that is only used once like this. Two reasons, firstly it makes the code harder to read because you have to go and find the value. Secondly the define has a much larger scope.

In any case i'm still not 100% on the numbers over 100 thing. I was just thinking to remove future conflicts but it also prevents moving existing out of tree modes into lua with the same number.

Reply:
Magic number is gone.